### ISSUE TITLE
map on String could potentially be faster

### ISSUE THREAD

**Issue by KristofferC** [JuliaLang/julia#32687]
Map on String uses the generic `AbstractString` interface but could potentially be faster by writing a custom version for `String` with something like:

```jl
function map2(f, s::String)
    out = Base.StringVector(max(4, sizeof(s)))
    offset = 1
    for c in s
        c′ = f(c)
        isa(c′, Char) || throw(ArgumentError(
            "map(f, s::String) requires f to return Char; " *
            "try map(f, collect(s)) or a comprehension instead"))
        offset += Base.__unsafe_string!(out, c′, offset)
        offset + 4 > length(out) && resize!(out, 2 * length(out))
    end
    resize!(out, offset-1)
    sizehint!(out, offset-1)
    return String(out)
end
```

with benchmarks like

```jl
julia> @btime map(uppercase, r);
  13.422 ms (5 allocations: 976.91 KiB)

julia> @btime map2(uppercase, r);
  6.947 ms (3 allocations: 2.86 MiB)

julia> @btime map(x->'0', r);
  8.175 ms (5 allocations: 976.91 KiB)

julia> @btime map2(x->'0', r);
  1.305 ms (3 allocations: 2.86 MiB)
```

**Comment by StefanKarpinski** [comment]
I've assigned myself since I want to make sure this happens, but if someone else wants to tackle it feel free and ping me for advice.

**Comment by stevengj** [comment]
It seems like that implementation could be used for the `::AbstractString` implementation, since it [always returns a `String`](https://github.com/JuliaLang/julia/blob/442d1599e62b1a3a453bd0c28c9a795e09e939ff/base/strings/basic.jl#L577) regardless of the input-string type.   Just replace `sizeof(s)` with `s isa String ? sizeof(s) : 64)` or similar.

**Comment by ps-pat** [comment]
@StefanKarpinski I'm new to this and I'd like to help!

**Comment by stevengj** [comment]
@ps-pat, just create a pull request altering [the `map` implementation for strings](https://kshyatt.github.io/post/firstjuliapr/) along the lines discussed above.   See https://kshyatt.github.io/post/firstjuliapr/  on creating a pull request if you aren't familiar with that.

**Comment by vtjnash** [comment]
Implemented by #33989
